odl/operator/operator_utilities.py

```diff
@@ -0,0 +1,114 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Usefull utility functions on discrete spaces (i.e., either Rn/Cn or
+discretized function spaces), for example obtaining a matrix representation of
+an operator. """
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+# External
+import numpy as np
+
+# Internal
+from odl.space.base_ntuples import FnBase
+from odl.set.pspace import ProductSpace
+
+
+def matrix_representation(op):
+    """Returns a matrix representation of a linear operator.
+
+    Parameters
+    ----------
+    op : :class:`~odl.Operator`
+        The linear operator of which one wants a matrix representation.
+
+    Returns
+    ----------
+    matrix : `numpy.ndarray`
+        The matrix representation of the operator.
+
+    Notes
+    ----------
+    The algorithm works by letting the operator act on all unit vectors, and
+    stacking the output as a matrix.
+
+    """
+
+    if not op.is_linear:
+        print('WARNING: The operator is not linear; cannot produce matrix',
+              'representation of it.')
+        return
+
+    if not (isinstance(op.domain, FnBase) or
+            isinstance(op.domain, ProductSpace)):
+        print('WARNING: The operator domain is not discrete or produc space;',
+              'cannot produce matrix representation of it.')
+        return
+
+    if not (isinstance(op.range, FnBase) or
+            isinstance(op.range, ProductSpace)):
+        print('WARNING: The operator range is not discrete; cannot produce',
+              'matrix representation of it.')
+
+    # Get the size of the range, and handle ProductSpace
+    op_ran = op.range
+    op_ran_is_prod_space = isinstance(op_ran, ProductSpace)
+    if op_ran_is_prod_space:
+        num_ran = op_ran.size
+        n = [op_ran[i].size for i in range(num_ran)]
+    else:
+        num_ran = 1
+        n = [op_ran.size]
+
+    # Get the size of the domain, and handle ProductSpace
+    op_dom = op.domain
+    op_dom_is_prod_space = isinstance(op_dom, ProductSpace)
+    if op_dom_is_prod_space:
+        num_dom = op_dom.size
+        m = [op_dom[i].size for i in range(num_dom)]
+    else:
+        num_dom = 1
+        m = [op_dom.size]
+
+    # Generate the matrix
+    matrix = np.zeros([np.sum(n), np.sum(m)])
+    tmp_1 = op_ran.element()
+    v = op_dom.element()
+    index = 0
+
+    for i in range(num_dom):
+        for j in range(m[i]):
+            v.set_zero()
+            if op_dom_is_prod_space:
+                v[i][j] = 1.0
+            else:
+                v[j] = 1.0
+            tmp_2 = op(v, out=tmp_1)
+            tmp_3 = []
+            if op_ran_is_prod_space:
+                for k in range(num_ran):
+                    tmp_3 = np.concatenate((tmp_3, tmp_2[k].asarray()))
```

Comment:
This is very inefficient. `np.concatenate` is really only a last resort when you don't know beforehand or it is very hard to calculate how large your final array will be. Otherwise, it's always faster to allocate once and then fill the array by slicing. Quick speed test:
```python
>>> x = np.ones(5000)
>>> y = np.empty(500000)
>>> def assign(x, y):
...     for i in range(100):
...         y[i * 5000: (i + 1) * 5000] = x   # Slicing -> fast and memory-friendly

>>> def concat1(x):
...     y = []
...     for i in range(100):
...         y = np.concatenate((y, x))  # Your version
        
>>> def concat2(x):
...     lst = [x] * 100   # Need to store all arrays -> bad
...     y = np.concatenate(lst)  # Fast, only one call

>>> %timeit assign(x, y)
1000 loops, best of 3: 214 µs per loop

>>> %timeit concat1(x)
10 loops, best of 3: 17 ms per loop

>>> %timeit concat2(x)
1000 loops, best of 3: 191 µs per loop